Increase default keepAliveTimeout to 60s and warn on legacy 5s usage

pras529 opened the pull request:
### Summary

This PR increases the default value of `http.Server.keepAliveTimeout` from **5000ms (5s)** to **60000ms (60s)**. This change aligns Node.js with the keep-alive timeout expectations of most browsers, proxies, and mobile clients, reducing the risk of subtle long-lived connection failures due to premature socket closure.

### Details

- **Default Change:**  
  The `keepAliveTimeout` default is now set to 60 seconds (60000ms) instead of the previous 5 seconds (5000ms).
- **Warning on Legacy Default:**  
  If a user explicitly sets `keepAliveTimeout` to `5000`, a warning is emitted highlighting the risks and recommending a higher value.
- **Warning on Default Change:**  
  If the default is used (i.e., not set by the user), a warning is emitted notifying the application about the new default and referencing the related Node.js issue ([#59193](https://github.com/nodejs/node/issues/59193)).
- **Documentation & Comments:**  
  All relevant documentation and code comments have been updated to reflect the new default.

### Motivation

The previous default of 5 seconds is shorter than the defaults used by common infrastructure, which can lead to unexpected connection terminations or failures. Setting the default to 60 seconds brings Node.js in line with industry standards and improves interoperability.

### Breaking Change

This is a breaking change as it may affect applications that implicitly relied on the old default. Users who require the old behavior should explicitly set `keepAliveTimeout` to `5000` in their code.

### References

- [Node.js issue #59193](https://github.com/nodejs/node/issues/59193)
- [Relevant discussion on keep-alive timeouts](https://github.com/nodejs/node/issues/59193)

---
**Please review carefully as this changes default server behavior for all applications that do not explicitly set `keepAliveTimeout`.**

pras529 commented:
@jasnell @targos @lpinca Please review my pull request when you have a chance. Thank you!

pras529 commented on doc/changelogs/CHANGELOG_V24.md:
```diff
@@ -624,6 +624,10 @@ latest release and evaluate their potential impact on your applications.
 ## Notable Changes
 
+### Breaking Changes
+- **http:** Changed the default value of `http.Server.keepAliveTimeout` from **5000ms (5s)** to **60000ms (60s)**. This prevents subtle long-connection failures due to premature socket closure, aligning with industry norms (browsers, proxies, mobile clients). See [issue #59193](https://github.com/nodejs/node/issues/59193).
```
I've resolved this issue and removed the breaking change alert regarding http.Server.keepAliveTimeout. The documentation now reflects the correct status, in line with the release team's guidance.

pras529 commented:
@mcollina , I've resolved this issue and removed the breaking change alert regarding http.Server.keepAliveTimeout. The documentation now reflects the correct status, in line with the release team's guidance. Please review my pull request when you have a chance. Thank you!

mcollina commented:
I've seen 60 seconds cause problems due to time network delays. Going over 60 seconds is necessary to account for that.

mcollina commented:
I literally picked 72 with no reasoning. We could do 65.

pras529 commented:
Thanks for clarifying. I’m fine with 65 seconds if that works better—happy to update the default to 65 to provide a bit more buffer over 60. Let me know if you have any concerns or if you think another value might be preferable.

pras529 commented:
@mcollina Yes, I’ll address the failing tests and linting issues—just need a bit of time to work through them.

pras529 commented on lib/_http_server.js:
```diff
@@ -481,7 +481,7 @@ function storeHTTPOptions(options) {
     this.keepAliveTimeout = keepAliveTimeout;
   } else {
-    this.keepAliveTimeout = 5_000; // 5 seconds;
+    this.keepAliveTimeout = 60000; // 60 seconds;
```
Sure! @pimterry

pimterry reviewed (approved):
Code all looks good to me :+1:

The first commit message is failing the linting - you'll need to use git to modify that. It's probably best to just squash all your commits and change the resulting commit message to describe the end result of the whole change here, because that message is what will be shown in the CHANGELOG when this is released.

pras529 commented:
Hello @mcollina,

Thank you for your review. I’ve addressed the changes you requested and updated the commit message to fully comply with our guidelines. I’m still a bit uncertain about the Linux and macOS test results—could you please take a look and let me know if anything else needs adjustment?

Thanks for your help!

mcollina commented on doc/api/http.md:
```diff
@@ -1936,9 +1936,13 @@ value only affects new connections to the server, not any existing connections.
 <!-- YAML
 added: v8.0.0
+changes:
+  - version: v24.0.0
```
```suggestion
  - version: REPLACEME
```

mcollina reviewed (changes requested):
one more nit, sorry.

benjamingr reviewed (approved):
I think 65 is a much more reasonable default than 5. Apache are the only ones I found to reduce it and that was with a given threading model (open idle connections take a thread and are more expensive) and time.

The nginx 75s keepalive timeout sounds a lot more reasonable to me ( https://nginx.org/en/docs/http/ngx_http_core_module.html )

Especially since reverse proxying Node behind nginx is common so I suspect gateway timeouts are needlessly common now for people using node:http directly.

mcollina commented:
@benjamingr any opinion on semversiness?

BethGriggs reviewed (approved):
Just a quick +1 to raising the default - I’ve recently had to add support for overriding server config in a framework to bump this value. We’ve been using 65s as a default, but would defer to SMEs as to what’s most appropriate.

mcollina commented:
@pras529 can you adjust the first commit message? (You can also squash all commits too)

Ethan-Arrowood commented:
I'm content with 65 seconds. Appreciate this change greatly.

pimterry commented:
> What's the reason for it be semver-minor and not semver-major?

https://github.com/nodejs/node/pull/59203#issuecomment-3124276867

If we think there's a real risk that this could break anything in existing code, or if people would just prefer to be cautious, I'm not opposed to making this major, but personally I'd be quite surprised it caused any issues in practice. For users it should really just be a bug fix & performance improvement.

Do we have any specific likely scenario where this might break things?

mcollina commented:
@RafaelGSSy assumption is that risk in shipping on Node 24 before it goes LTS is close to zero, but very high afterwars.

ex1st commented:
Hi folks! If you plan to increase `keepAliveTimeout` to 65 seconds, should you also increase the `headersTimeout` to 66 at least? Please check https://github.com/nodejs/node/issues/27363

pimterry commented:
A very good question @ex1st! From what I can see though, that issue was resolved by https://github.com/nodejs/node/pull/32329/, which changed how headerTimeout was measured entirely, so it shouldn't overlap with keepAliveTimeout as it used to. From my reading of the code there, nowadays the header timer doesn't start until a new request starts, and it's disabled when it finishes, so it's not measuring idle time at all. Does that make sense?

pimterry commented:
It would be great to get this merged, @BethGriggs mentioned yesterday that it's still causing issues, and it would be a clear improvement for lots of use cases. It looks like there's a pending issue on the commit message and some debate about whether it's semver-major or not.

I'm happy to go for semver-major if there's uncertainty, just in case. Any objections to that? I'll update it now but shout if anybody feels that's an issue. That'll delay releasing this significantly unfortunately but it's better than leaving it stuck.

@pras529 can you update the commit message & force push, so the checks pass? Since this was done the requirements have also changed, so you'll now also need to include a 'Signed-Off-By' footer - see the docs at https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md#commit-message-guidelines for details & examples. En route, a rebase would be helpful since this is now a long way back from `main`, and the tests will need to be passing too, if that doesn't get resolved by the rebase & rerun (logs are gone, so it's unclear what failed there).

RafaelGSS commented:
@pimterry feel free to open a fresh new PR and add @pras529 on Co-Authored-By section in case of silent communication.

pimterry commented:
> feel free to open a fresh new PR and add @pras529 on Co-Authored-By section in case of silent communication.

Now done - https://github.com/nodejs/node/pull/62782 has these changes, but rebased, with the linting & tests passing. I'll close this and we can continue over there. 

Reviews welcome please from anybody here who's still following this :smile:
